### PR TITLE
fix(replay): reporting and lifecycle cleanup

### PR DESCRIPTION
**Pull request by jasonqinzhou**
## Why and what changed

This PR is scoped to **replay reporting and request-lifecycle cleanup**. Large offline replays retain completed request state and allocate additional latency arrays at report time. The change releases state after its final consumers are done and stores exact summary latency samples on disk when their in-memory buffer fills.

The implementation covers three related parts:

1. **Reporting cleanup:** fold finished requests into aggregate statistics and spill exact latency samples to replay-owned anonymous temporary files after 4,096 samples per distribution. All samples contribute to the report; counts and rounded-rank percentiles remain exact. Mean/std can differ at floating-point roundoff.
2. **Lifecycle cleanup:** retire completed batch-summary records after completion callbacks, and release disaggregated request state after both pipelines and the handoff coordinator are quiescent. A cancellation-accounting guard ensures that the same request is retired exactly once.
3. **Supporting reporting correctness:** configure SLA thresholds before incremental aggregation, propagate actual temporary-storage failures through Rust/Python, and cover these paths with parity, lifecycle, and boundary tests.

Active requests and steppable SDK completed-request queries retain their existing behavior. Detailed reports keep their complete, ordered in-memory list API. Temporary files are released when reporting finishes or fails; existing user files are untouched.

Scope excludes workload caps, host-budget enforcement, sampling, truncation, a new streaming-output API, and changes to performance models or search policy. The 4,096-sample threshold only chooses storage location. Temporary-disk use grows with the workload, and actual storage failures stop the replay rather than produce a successful partial report.

## Review map

- Risk level: medium (request lifecycle and reporting).
- Scope audit: all 12 changed files at `cc7ddf09810e848d5b3c8989550074224c6aee3f` belong to the reporting/lifecycle implementation, its cross-language error propagation, tests, or documentation. No unrelated change was identified.

| Files | Purpose within this scope |
| --- | --- |
| `crates/core/src/replay/report.rs`, `report/bounded.rs` | Completed-record retirement, incremental statistics, exact disk-backed latency samples, and preservation regressions. |
| `crates/core/src/replay/agg.rs`, `disagg.rs`, `replayer.rs` | Apply cleanup only to batch execution, wait for lifecycle completion, and initialize SLA thresholds before incremental reporting. |
| `crates/core/src/engine/scheduler/rank.rs` | Guard against publishing duplicate retirement for a canceled request; this is a lifecycle-accounting fix. |
| `crates/core/src/replay/error.rs`, `crates/core/src/python.rs`, `python/aisimulate/src/aisimulate/runner.py` | Carry actual reporting-storage failures through the existing execution interfaces. They do not enforce a resource budget. |
| `crates/core/src/replay/disagg_tests.rs`, `tests/test_runner.py`, `crates/core/src/replay/README.md` | Lifecycle/error-boundary regressions and documentation of the cleanup behavior. |

- Public or serialized contract: successful report schemas and detailed-record ordering are preserved. The Rust error enum adds `ResourceLimited`, mapped to Python `MemoryError` and to the host-resource exception when that optional API is installed, for actual reporting-storage failures.
- Compatibility: active state, detailed/canonical capture, and steppable request queries retain the state their consumers need. Detailed records, active work, input traces, and trajectory metadata still require host memory.

## Evidence

Refreshed against main `c0317fccb25385093f9e8948675f765e4a6a0e96`, resolving reporting/replayer/test conflicts while preserving the current report contract. Runtime implementation validated on `830b3d33a39f0837d7a911034bfc0208f3b3510a`. Final head `cc7ddf09810e848d5b3c8989550074224c6aee3f` also preserves typed storage failures during final draining, adds native error-mapping and write-failure regressions, and corrects the incremental-reporting documentation:

- `cargo test -p aisimulate-core --lib replay:: --locked --offline -- --quiet`: 310 passed on the final head.
- `cargo test -p aisimulate-core --lib --locked --offline -- --quiet`: 1,397 passed, 1 ignored (includes scheduler coverage).
- `PYTHONPATH=python/aisimulate/src pytest -c python/aisimulate/pytest.ini -p no:timeout tests/test_runner.py -q`: 107 passed (4 warnings).
- `cargo check -p aisimulate-core --features python --locked --offline` with `PYO3_PYTHON` set to Python 3.12: passed.
- Rust formatting, Ruff lint/format, and `git diff --check`: passed.
- Native error-mapping regression: `cargo test -p aisimulate-core --features embed-python --lib python::tests::replay_python_error --locked --offline -- --nocapture` with Python 3.12: 2 passed. Passed on `bc7ec1d1`; these tests and the binding are unchanged in the final head. This covers contextual reporting-storage errors mapping to `MemoryError`, unchanged ordinary-error mapping to `RuntimeError`, and complete error messages.
- Storage-failure regression: a real read-only file makes writes fail in both completed-record retirement and final draining. The latter failed before the correction; both now return `ResourceLimited` without preparing a partial report.
- Boundary cases: all 100,001 detailed records survive with their original ordering; a sparse-file regression appends beyond the former 256 MiB per-distribution cap without overwriting the existing tail; active requests and steppable queries survive completed-state retirement; truncated sample storage fails explicitly.
- Expected values: spilled percentiles are compared with the existing exact selection implementation. Summary/detail parity covers both backends, aggregated/disaggregated execution, SLA statistics, and simulation cutoffs.
- Native Python extension smoke test on final head `cc7ddf09`: 4,200 requests per run exercise real disk spilling; summary and detailed reports agree for vLLM/SGLang and aggregated/disaggregated execution after excluding wall-clock timing and its derived rates (numeric tolerance 1e-10). Setting `TMPDIR` to a missing directory lets a 16-request in-memory report succeed and makes spilled reports raise `MemoryError` with the underlying file-creation error. Both spaced arrivals (spill during execution) and simultaneous arrivals (spill at finalization) passed; no partial report is returned.
- [Fast CI](https://github.com/ai-dynamo/aisimulate/actions/runs/35291208139): passed on the final head. [Full CI](https://github.com/ai-dynamo/aisimulate/actions/runs/35291559889): completed with 51 successful jobs, but both application-contract shards failed on the inherited `test_nightly_versions_are_unique_date_ordered_and_stable_across_retries` (missing Node in Python-only containers), so `Full CI Success` failed. Each affected shard otherwise passed 1,888 tests and 132 subtests, with 5 expected skips. The unchanged nightly test is already fixed by separate [PR #259](https://github.com/ai-dynamo/aisimulate/pull/259).
- [CodeRabbit](https://github.com/ai-dynamo/aisimulate/pull/173#issuecomment-5627416826): final-head review completed with no actionable comments (run `cf164345-db57-4295-99be-8a70c899e9e8`, coverage explicitly identifies `cc7ddf09810e848d5b3c8989550074224c6aee3f`). All review threads are resolved. Human/CODEOWNER approval remains outstanding.
- Codex reviewed commit: `cc7ddf09810e848d5b3c8989550074224c6aee3f`.

**Readiness:** implementation review and scoped fixes are complete; human-review readiness remains blocked by Full CI. After #259 lands, refresh this branch and rerun the exact-head review/CI gates. The unrelated nightly CI fix is kept outside this reporting/lifecycle PR.

## Modeling or data provenance

No performance-model or measurement-data changes. Existing ITL/per-user-throughput sketches retain their existing 0.1% relative quantile error; no new approximation is introduced.

## Tracking

- Tracks [AIC-1955](https://linear.app/nvidia/issue/AIC-1955).
- [Dynamo #14679](https://github.com/ai-dynamo/dynamo/pull/14679) is related integration work; its core pin must be refreshed and validated before consuming this head.
- [AISimulate #174](https://github.com/ai-dynamo/aisimulate/pull/174) owns integrated Mac/Linux qualification. This PR does not establish full-scale completion or the original panic's root cause.

### CODE REVIEW

File: crates/core/src/replay/report.rs [hunk 1]
```diff
@@ -864,6 +870,71 @@ impl TraceRequestStats {
...
+    ) -> anyhow::Result<()> {
+        self.retire_completed()
+            .map_err(|error| anyhow::anyhow!("resource_limited: report storage: {error:#}"))?;
+        if self.batch_reporting
+            && self.bounded_summary.is_none()
+            && self.requests.len() >= bounded::MAX_RETAINED_REQUESTS
+        {
+            anyhow::bail!(
+                "resource_limited: detailed replay reporting exceeds the 100000 retained-request limit; use summary output or a smaller explicit workload"
+            );
```
Comment (jasonqinzhou) [inline crates/core/src/replay/report.rs]: The report-size and spool failures currently return untyped anyhow errors. `runtime_error` classifies those as `ReplayError::Invariant`, so the Python/sweeper boundary can treat host resource exhaustion as a failed simulation. Preserve a typed resource-limit error through the Rust and native Python boundaries, and translate it to the host resource exception when that API is installed.

**Comment by jasonqinzhou** [comment]
Post-publication review follow-up: the non-test Python-feature build warns that `SlaThresholds` is unused in `replayer.rs:34` after moving SLA setup into the runtime. Move the import into the generated-replay test module so strict Rust lint builds remain clean. The unchanged import line cannot accept a diff comment.

File: crates/core/src/replay/report.rs (hunk 1 above)
Comment (jasonqinzhou) [reply crates/core/src/replay/report.rs]: Fixed in a3ffdf97e28356769ad5e4194caf31d3f6a768c0: typed ReplayError::ResourceLimited survives the Rust boundary, the native layer emits MemoryError, and resource-aware Python runners translate it to ResourceLimitError. Replay tests and the Python boundary regression pass.

File: crates/core/src/replay/report/bounded.rs [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+//! Temporary files are anonymous and removed on success, error, or process exit.
+use std::fs::File;
+use std::io::{BufReader, BufWriter, Read, Seek, SeekFrom, Write};
+
+use super::*;
+use anyhow::{Context, Result, bail};
+
+const MEMORY_SAMPLES: usize = 4096;
+// Four distributions together consume at most 1 GiB of temporary disk space.
+const MAX_SAMPLES: usize = (256 * 1024 * 1024) / 8;
```
Comment (jasonqinzhou) [inline crates/core/src/replay/report/bounded.rs]: The cleanup-only scope agreed for this PR should preserve the complete workload and detailed-report API. The fixed 1 GiB sample cap and 100,000-request guard can refuse otherwise valid runs even when the host has capacity. Remove these policy limits while retaining exact spill-to-disk statistics, post-callback completed-state retirement, and explicit handling of actual storage failures. Add regressions for crossing the previous detailed-record and sample-count boundaries without truncation.

Comment (jasonqinzhou) [reply crates/core/src/replay/report/bounded.rs]: Fixed in c3bf647d6df554fcc17171d8915c8ba972f234a5. Removed the fixed sample-storage and detailed-request caps while preserving exact spill-to-disk statistics and the complete ordered detailed-report API. Added regressions for 100,001 complete detailed records, appending past the previous sample-spool boundary, and retaining active state plus steppable completed-request queries. Local validation: 300 replay tests, 13 scheduler tests, 88 Python runner tests, Python-feature Rust check, formatting and lint passed. The PR remains a draft; hosted checks and human review are still separate gates.

**Comment by jasonqinzhou** [comment]
Readiness review of c3bf647d found conflicts with current main c0317fccb25385093f9e8948675f765e4a6a0e96 in replay/replayer.rs, replay/report.rs, and tests/test_runner.py. Resolve these against the current reporting contract and repeat summary/detail parity and lifecycle validation before claiming human-review readiness. The scope remains reporting and request-lifecycle cleanup.

**Review by jasonqinzhou — COMMENTED** [review]
Codex review of exact head 830b3d33a39f0837d7a911034bfc0208f3b3510a against main c0317fccb25385093f9e8948675f765e4a6a0e96: the 12-file diff matches reporting and request-lifecycle cleanup. Traced terminal callbacks, disaggregated quiescence/cancellation, summary aggregation and SLA setup, exact sample spilling, detailed/steppable retention, current report metadata, and Rust/Python storage-error propagation. No additional actionable finding remains from this pass. Resolved the current-main conflicts while preserving the existing reporting fields. Local validation: 1,397 Rust library tests passed, 1 ignored; the 309 replay tests also passed separately; 107 Python runner tests passed; Python-feature Rust check, Rust formatting, Ruff and whitespace checks passed. Fast CI, substantive CodeRabbit review and Full CI still require verification on this head before a human-review-ready claim.

File: tests/test_runner.py [hunk 3]
```diff
@@ -1254,6 +1254,25 @@ def test_memory_detail_with_explicit_blocks_does_not_guess_components():
...
+    class HostResourceError(RuntimeError):
+        pass
+
+    resources = ModuleType("aisimulate.resources")
+    resources.ResourceLimitError = HostResourceError
+    monkeypatch.setitem(sys.modules, "aisimulate.resources", resources)
+
+    class LimitedRuntime:
+        def run_replay_json(self, payload):
+            raise MemoryError("report storage: No space left on device")
```
[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply tests/test_runner.py:1270]: Confirmed the permanent regression gap. A native-extension smoke test on 830b3d33 already exercised an actual temporary-file failure and received MemoryError through the Rust binding. I will retain that evidence and the host-side Python test, and add a Rust regression for contextual ResourceLimited mapping plus the ordinary-error fallback.

[the remark this answers is not in the document]
Comment (jasonqinzhou) [reply tests/test_runner.py:1270]: Fixed in bc7ec1d18e8adc036b34536cbed32ead45cde967. Added Rust binding regressions for ResourceLimited under two context wrappers -> PyMemoryError with the complete message, and contextual ordinary errors -> PyRuntimeError. Both passed with cargo test -p aisimulate-core --features embed-python --lib python::tests::replay_python_error (Python 3.12). The existing host-side conversion test remains unchanged.

**Review by jasonqinzhou — COMMENTED** [review]
Codex review of exact head bc7ec1d18e8adc036b34536cbed32ead45cde967 against main c0317fccb25385093f9e8948675f765e4a6a0e96: clean. Rechecked the complete 12-file reporting/lifecycle diff; the only change since the prior substantive review is two native error-mapping regression tests. Both pass with embed-python and Python 3.12, covering contextual ResourceLimited -> PyMemoryError, the PyRuntimeError fallback, and preserved messages. The runtime implementation previously passed 1,397 Rust library tests (1 ignored), 107 Python runner tests, and a native extension audit comparing spilled/detailed reports for 4,200 requests across both backends/topologies and exercising a real temporary-file creation failure. No actionable findings remain from this review. Final-head Fast CI, CodeRabbit, and Full CI must complete before human-review readiness is claimed.

File: crates/core/src/replay/report.rs [hunk 4]
```diff
@@ -1063,6 +1069,67 @@ impl TraceRequestStats {
...
+            return Ok(());
+        };
+        for (_, mut stats) in self.requests.drain() {
+            stats.finalize_token_timeline(
+                stats.terminal_status == Some(ReplayTerminalStatus::Completed)
+                    && stats.first_admit_ms.is_some(),
+                &mut self.itl_distribution,
+                &mut self.output_token_throughput_per_user,
+            );
+            summary.add(&stats, self.sla)?;
```
Comment (jasonqinzhou) [inline crates/core/src/replay/report.rs]: The final-drain add is the one reporting-storage call that propagates an unclassified anyhow error. Fresh batch runs retire terminal records above, but a collector entering batch mode with retained completed records takes this branch. Apply the same ResourceLimited conversion here and fault-inject a write failure into both the retirement and final-drain paths to preserve a consistent error boundary.

**Review by jasonqinzhou — COMMENTED** [review]
Validated the auxiliary-review documentation issue and a reporting-error consistency gap. Fresh batch runs queue completed records before final draining, so the claimed normal-run final-drain failure is not established; nevertheless every storage-writing branch in prepare_batch_report should preserve the same typed error contract.

**Comment by jasonqinzhou** [comment]
The Metrics Collection paragraph in crates/core/src/replay/README.md still says the harness does not aggregate incrementally. That contradicts this PR. I will update it to distinguish incremental batch-summary collection from detailed/steppable report construction. Auxiliary bot references to merge conflicts and an unused SlaThresholds import are stale: the conflicts were resolved in 830b3d33, and SlaThresholds is imported only inside the generated-replay test module.

File: crates/core/src/replay/report.rs (hunk 4 above)
Comment (jasonqinzhou) [reply crates/core/src/replay/report.rs]: Fixed in cc7ddf09810e848d5b3c8989550074224c6aee3f. The final-drain write now uses the same ResourceLimited conversion as normal retirement and report finalization. The fault-injection regression uses a real read-only file and covers both retirement and pre-retained completed-state draining; it failed on the latter before the fix and now passes. All 310 replay tests pass. The same commit corrects the Metrics Collection documentation.

**Review by jasonqinzhou — COMMENTED** [review]
Codex review of exact head cc7ddf09810e848d5b3c8989550074224c6aee3f against main c0317fccb25385093f9e8948675f765e4a6a0e96: no remaining actionable finding. Re-read the complete 12-file diff, including lifecycle callback/quiescence ordering, complete detailed and steppable data retention, exact spilled statistics, early SLA configuration, cross-language error mapping, and documentation. Confirmed the final-drain error conversion with a real write-failure regression that failed before the fix and now passes in both retirement paths; all 310 replay tests pass. Native mapping regressions passed in the preceding test-only commit and are unchanged. Earlier runtime validation (1,397 Rust library tests, 107 Python runner tests, native spilled-summary parity and real file-creation failure) is recorded separately in the description. Scope remains reporting and lifecycle cleanup, with no workload cap, sampling, or model/search-policy change. Final-head CodeRabbit and hosted CI are still pending.

**Comment by jasonqinzhou** [comment]
Review closeout for cc7ddf09810e848d5b3c8989550074224c6aee3f: the complete 12-file diff matches reporting and lifecycle cleanup. Resolved current-base conflicts, added native error-mapping regressions, preserved the typed storage error in final draining, and corrected stale reporting documentation. All actionable threads are resolved; CodeRabbit completed its final-head review without actionable comments and Fast CI passed.

Full CI is terminal but not green: 51 jobs passed, including both Rust architectures, public APIs, feature modes, engine goldens, prediction regression, platform wheels and the other application shards. Both application-contract shards failed only on test_nightly_versions_are_unique_date_ordered_and_stable_across_retries because shutil.which("node") returns None in Python-only containers (1,888 tests and 132 subtests otherwise passed per shard, 5 skips). These two nightly-test functions are unchanged from current main. Existing PR #259 addresses exactly this failure by running JavaScript behavior in the Node suite while retaining Python PEP 440 checks.

CI run: https://github.com/ai-dynamo/aisimulate/actions/runs/35291559889
Dependency: https://github.com/ai-dynamo/aisimulate/pull/259

This is not yet human-review-ready under REVIEW.md. Keep the CI repair in #259 to preserve this PR's requested scope; after it lands, refresh #173 and rerun exact-head review and CI. No merge or approval has been performed.

**Review by simone-chen — APPROVED** [review]
